### DISCUSSION TITLE
Hard or soft constraints.

### DISCUSSION THREAD

**Discussion by Bestbbb** [google/or-tools#3640] [CP-SAT questions] [answered]
How can I implement hard or soft constraints,now I use such as addGreaterThan seems to be hard constraints,if I want to use some soft constrains,how can I achieve it.

**Comment by jmarca** [comment]
Which solver?

**Reply by Bestbbb** [reply]
cpsolver

**Comment by lperron** [comment, marked as answer]
2 options:

add slack: model.Add(x > 0 - slack) and minimize slack (a positive variable)
use enforcement literals: model.Add(x > 0).OnlyEnforceIf(bool_var), and maximize(sum(bool_vars))
